### PR TITLE
Implementation of contraints

### PR DESCRIPTION
**Pull request by lijas**
Constraints, very important

### CODE REVIEW

**Comment by KristofferC** [comment]
?

**Comment by lijas** [comment]
Såg ett fel och ville fixa det och trodde man var tvungen att stänga pr....
Inte så bra på github...

**Comment by KristofferC** [comment]
Aha, ne. Bara att pusha fler commits till denna branch så uppdateras det.

File: src/Dofs/Constraints.jl [hunk 1]
```diff
@@ -0,0 +1,116 @@
+export Constraint, Constraints, add!, nconstraints, applypenelty
```
Comment (KristofferC) [inline src/Dofs/Constraints.jl]: `penelty` -> `penalty`

File: src/Dofs/Constraints.jl [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+"""
+    Constraint
+Takes care of the constraints in the system. Applies the contributions 
+from the constraint the stiffness matrix.
```
Comment (KristofferC) [inline src/Dofs/Constraints.jl]: "to the stiffness matrix"

File: src/Dofs/Constraints.jl [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+Important functions:
+
+cons = Constraints(dh::DofHandler)
+add!(cons, function, node(s))
+...
+applypenelty(cons, K, Symmetric}, f, penelty, X)
```
Comment (KristofferC) [inline src/Dofs/Constraints.jl]: Denna verkar konstig med `Symmetric}`

File: src/Dofs/Constraints.jl [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+function applypenelty!(cons::Constraints, K::Union{SparseMatrixCSC, Symmetric}, f::AbstractVector, penelty::Float64, X::AbstractVector)
+    
+    ncons_eq = nconstraints(cons);
+    if ncons_eq == 0
+       return  
+    end
+	
+    constraints = cons.constraints;
+
+    rowHess = Int[]; colHess=Int[]; dataHess=Float64[];
```
Comment (KristofferC) [inline src/Dofs/Constraints.jl]: image

File: src/Dofs/Constraints.jl (hunk 1 above)
Comment (lijas) [inline src/Dofs/Constraints.jl]: Stavning är min starka sida

File: src/Dofs/Constraints.jl [hunk 5]
```diff
@@ -105,7 +105,7 @@ function applypenelty!(cons::Constraints, K::Union{SparseMatrixCSC, Symmetric},
     end
 
     G = sparse(rowGrad,colGrad,dataGrad, ncons_eq, ndofs(cons.dh))
-    H = sparse(rowHess,colHess,dataHess, ndofs(cons.dh), ndofs(cons.dh))
+    H = sparse(rOwHeSs,colHess,dataHess, ndofs(cons.dh), ndofs(cons.dh))
```
Comment (KristofferC) [inline src/Dofs/Constraints.jl]: lol!

**Comment by lijas** [comment]
I lost the changes in this PR because I was a noob at git. Closing

**Comment by KristofferC** [comment]
I'm pretty sure you can retrieve them somehow (if you want to). Check out `git reflog`. You can go to any old commit there with `git checkout COMMIT_SHA` etc.

**Comment by lijas** [comment]
I saved the files on my computer, so I might make a new PR sometime

**Comment by KristofferC** [comment]
This is pretty good reading regarding constraints btw https://www.dealii.org/8.4.1/doxygen/deal.II/group__constraints.html

**Pull request closed without merging**